### PR TITLE
Feature/undo - Make it possible to undo the last answered question

### PR DESCRIPTION
**Pull request by SebastianKapunkt**
- [x] tested manually

### CODE REVIEW

File: threesixty/templates/threesixty/answer_form.html [hunk 1]
```diff
@@ -8,11 +8,12 @@ <h2 class="w3-center">{{ name }}</h2>
     </div>
     <form method="POST" style="display: none">
         {% csrf_token %}
-        {{ form.as_p }}
+        {{ form }}
```
Comment (codingjoe) [inline threesixty/templates/threesixty/answer_form.html]: is this change related? Furthermore `p` style forms are better supported in W3CSS, `__str__` defaults to `as_table`.
I would ask you to please review this change.

File: threesixty/views.py [hunk 2]
```diff
@@ -260,21 +262,39 @@ def post(self, request, *args, **kwargs):
 
     def get(self, request, *args, **kwargs):
         self.participant = get_object_or_404(models.Participant, email=self.email, survey=self.survey)
+        if self.kwargs.get('question_pk'):
+            return self.get_specific_question(request, args, kwargs)
+        else:
+            return self.get_random_question(request, args, kwargs)
+
+    def get_specific_question(self, request, args, kwargs):
+        question_pk = self.kwargs['question_pk']
```
Comment (codingjoe) [inline threesixty/views.py]: You are getting it for the second time from the kwargs. Why not attach it to the instance in the get method.

File: threesixty/views.py [hunk 3]
```diff
@@ -260,21 +262,39 @@ def post(self, request, *args, **kwargs):
 
     def get(self, request, *args, **kwargs):
         self.participant = get_object_or_404(models.Participant, email=self.email, survey=self.survey)
+        if self.kwargs.get('question_pk'):
```
Comment (codingjoe) [inline threesixty/views.py]: Honestly I wouldn't try to solve two different things in one view.
I would suggest to use a regular `UpdateView` when you want to go to a specify question.

File: threesixty/views.py [hunk 4]
```diff
@@ -1,18 +1,20 @@
...
 from django.db import connection
-from django.http import Http404, HttpResponseRedirect, JsonResponse, HttpResponseForbidden
 from django.db.models import Avg, FloatField, IntegerField
 from django.shortcuts import get_object_or_404
 from django.template.loader import render_to_string
 from django.urls import reverse
 from django.views import generic
 
 from . import forms, models
+from .models import Answer, Question
```
Comment (codingjoe) [inline threesixty/views.py]: Please don't double import `models`

File: threesixty/views.py [hunk 5]
```diff
@@ -290,10 +310,31 @@ def get_initial(self):
...
+                latest_answer = Answer.objects.filter(participant=self.participant).latest('created')
+                kwargs = {
+                    'survey_pk': self.survey.pk,
+                    'token': self.token,
+                    'question_pk': latest_answer.question.pk
+                }
+                latest_answer.delete()
+                return HttpResponseRedirect(reverse('surver-answer-specific', kwargs=kwargs))
+            except Answer.DoesNotExist:
+                return self.redirect_survey_answer(self.survey.pk, self.token)
```
Comment (codingjoe) [inline threesixty/views.py]: Why? Who says that all questions are answered at this point? Why not go to `self.request.path`?

File: threesixty/views.py [hunk 6]
```diff
@@ -290,10 +310,31 @@ def get_initial(self):
         return initial
 
     def form_valid(self, form):
-        if not self.survey.participant_can_skip and form.cleaned_data['decision'] is None:
+        if self.survey.participant_can_skip is False and form.cleaned_data['decision'] is None:
```
Comment (codingjoe) [inline threesixty/views.py]: Good idea, but unrelated. Please put this into a separate PR.

File: threesixty/views.py (hunk 5 above)
Comment (SebastianKapunkt) [reply threesixty/views.py]: Its the other way around. There are no more answeres, that means no question was answered yet and thats why we redirect to a random question.

File: threesixty/views.py (hunk 2 above)
Comment (SebastianKapunkt) [reply threesixty/views.py]: I would not attach it to the instance, but just get it once.

File: threesixty/views.py (hunk 3 above)
Comment (SebastianKapunkt) [reply threesixty/views.py]: Would be nice if we can talk more in detail about this. I am not sure how to achieve that. As far as I can imagine it would mean there is another form that would post to another view when we want to undo?

**Comment by amureki** [comment]
@SebastianKapunkt great job here! 🥇
